src/storage/sparse_ann_query.rs

```diff
@@ -0,0 +1,97 @@
+use dashmap::DashMap;
+use rayon::iter::IntoParallelRefIterator;
+use rayon::prelude::*;
+use std::collections::HashMap;
+use std::{cmp::Ordering, collections::BinaryHeap};
+
+use crate::models::lazy_load::LazyItemVec;
+use crate::models::types::SparseVector;
+
+use super::inverted_index_sparse_ann::{InvertedIndexSparseAnn, InvertedIndexSparseAnnNode};
+
+const K: usize = 5;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct SparseAnnResult {
+    vector_id: u32,
+    similarity: f32,
+}
+
+impl Eq for SparseAnnResult {}
+
+impl PartialOrd for SparseAnnResult {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.similarity.partial_cmp(&other.similarity)
+    }
+}
+
+impl Ord for SparseAnnResult {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap_or(Ordering::Equal)
+    }
+}
+
+pub struct SparseAnnQuery {
+    /// Query vector is a pair of non-zero values and its dimension
+    query_vector: SparseVector,
+}
+
+impl SparseAnnQuery {
+    pub fn new(query_vector: SparseVector) -> Self {
+        SparseAnnQuery { query_vector }
+    }
+
+    pub fn sequential_search(&self, index: &InvertedIndexSparseAnn) -> Vec<SparseAnnResult> {
+        let mut dot_products: HashMap<u32, f32> = HashMap::new();
+
+        let mut sorted_query_dims: Vec<(u32, f32)> = self.query_vector.entries.clone();
+        sorted_query_dims.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(Ordering::Equal));
+
+        // Iterate over the query vector dimensions
+        for &(dim_index, dim_value) in &sorted_query_dims {
+            if let Some(node) = index.find_node(dim_index) {
+                let quantized_query_value = InvertedIndexSparseAnnNode::quantize(dim_value);
+                let start_key: u8 = 63u8;
+                let end_key: u8 = match quantized_query_value {
+                    0..=15 => 47,
+                    16..=31 => 31,
+                    32..=47 => 15,
+                    _ => 0,
+                };
+                for key in start_key..=end_key {
+                    let lazy_item_vec: &LazyItemVec<u32> = &node.shared_get().data[key as usize];
+                    if !lazy_item_vec.is_empty() {
+                        for lazy_item in lazy_item_vec.iter() {
+                            let vector_id = lazy_item.get_data(index.cache.clone());
+                            let dot_product = dot_products.entry(*vector_id).or_insert(0.0);
+                            *dot_product += (quantized_query_value * key) as f32;
+                        }
+                    }
+                }
+            }
+        }
+
+        // Create a min-heap to keep track of the top K results
+        let mut heap = BinaryHeap::with_capacity(K + 1);
```

Comment:
Let's abstract some of this common logic into common modules that are used in both kNN and aNN so that we can be assured that bugs are not introduced on duplication.